### PR TITLE
Split manpage generation into a Sphinx extension and a PEP 517 in-tree submodule with CLI

### PR DESCRIPTION
**Pull request by webknjaz**
##### SUMMARY

This patch attempts to get rid of the `generate-man` hacking script by moving the corresponding logic to a Sphinx extension and the CLI interface for making manpages to the PEP 517 build backend.

##### ISSUE TYPE

- Docs Pull Request
- Maintenance Pull Request
- Packaging Pull Request

##### COMPONENT NAME

Sphinx CLI pages, manpages, PEP 517 build backend.

##### ADDITIONAL INFORMATION

#80372

### CODE REVIEW

File: test/sanity/code-smell/rstcheck.py [hunk 1]
```diff
@@ -8,6 +8,13 @@
 
 def main():
     paths = sys.argv[1:] or sys.stdin.read().splitlines()
+    is_cli_manpageish = any(
+        _path.startswith('docs/docsite/rst/cli/ansible')
+        for _path in paths
+    )
+    optional_args = [
+        '--ignore-directives', 'ansible-cmd-docs'
+    ] if is_cli_manpageish else []
```
Comment (mattclay) [inline test/sanity/code-smell/rstcheck.py]: Split the paths into two lists and run `rstcheck` separately against each one, then concatenate the results. That will ensure each list of files is tested with the correct options.

File: lib/ansible/cli/console.py [hunk 2]
```diff
@@ -39,26 +39,30 @@ class ConsoleCLI(CLI, cmd.Cmd):
...
+       - ``become_user [user]``: set the become_user
+       - ``remote_user [user]``: set the remote_user
+       - ``become_method [method]``: set the privilege escalation method
+       - ``check [bool]``: toggle check mode
+       - ``diff [bool]``: toggle diff mode
+       - ``timeout [integer]``: set the timeout of tasks in seconds
+         (0 to disable)
+       - ``help [command/module]``: display documentation for
+         the command or module
+       - ``exit``: exit ``ansible-console``
```
Comment (mattclay) [inline lib/ansible/cli/console.py]: It looks like these changes could be split into a separate PR, correct?

**Review by mattclay — DISMISSED** [review]

File: test/sanity/code-smell/rstcheck.py (hunk 1 above)
Comment (webknjaz) [reply test/sanity/code-smell/rstcheck.py]: Great idea! Will do.

File: lib/ansible/cli/console.py (hunk 2 above)
Comment (webknjaz) [reply lib/ansible/cli/console.py]: Yes! I've made the commits atomic specifically for this reason :)

Comment (webknjaz) [reply lib/ansible/cli/console.py]: @mattclay I've extracted this into https://github.com/ansible/ansible/pull/80719

File: test/sanity/code-smell/rstcheck.py (hunk 1 above)
Comment (webknjaz) [reply test/sanity/code-smell/rstcheck.py]: @mattclay so... I ended up rewriting it heavily..

File: docs/docsite/_ext/cli_manpages/ext.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    """Reusable pre-loaded jinja2 template being rendered in the directive."""
+
+    def _make_doctree_cache_depend_on(self, *paths) -> None:
+        for path in paths:
+            self.env.note_dependency(str(path))
+
+    def run(self) -> t.List[nodes.Node]:
+        """Generate a node tree in place of the directive.
+
+        It deduces the CLI command, and the Python class and module names from
```
Comment (oraNod) [inline docs/docsite/_ext/cli_manpages/ext.py]:
```suggestion
        It determines the CLI command, and the Python class and module names from
```

File: docs/docsite/_ext/cli_manpages/ext.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        the current Sphinx document name. It also uses said Python module, this
+        extension module itself and the template for the Sphinx's built-in
```
Comment (oraNod) [inline docs/docsite/_ext/cli_manpages/ext.py]:
```suggestion
        extension module itself and the template for Sphinx's built-in
```

File: packaging/pep517_backend/_manpage_generation.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    jinja2_ctx = generate_cli_jinja2_context(cli_bin_name)
+    manpage_document = jinja2_template.render({
+        'ansible_core_version': ansible_version,
+        **jinja2_ctx,
+    })
+    return manpage_document
+
+
+def generate_manpages(target_dir: os.PathLike) -> None:
+    """Generate all manpages into given directory."""
```
Comment (oraNod) [inline packaging/pep517_backend/_manpage_generation.py]:
```suggestion
    """Generate all manpages into a given directory."""
```

File: test/sanity/code-smell/rstcheck.py [hunk 6]
```diff
@@ -4,59 +4,169 @@
...
+        )
+    except subprocess.CalledProcessError as proc_err:
+        raise RuntimeError(
+            'Oops... Got unexpected error output from '
+            f'``{rstlint_cmd_text !s}``:\n\n{proc_err.stderr !s}',
+        ) from proc_err
+
+    if rstlint_result.stdout:
+        raise RuntimeError(
+            f'Oops... Got unexpected output from ``{rstlint_cmd_text !s}``:'
```
Comment (oraNod) [inline test/sanity/code-smell/rstcheck.py:117]: Just pointing this out, take it or leave it but I'm not sure if "Oops..." is meaningful. It's fun and friendly but does it risk confusion, especially for non-native English speakers? IMO it's better to drop the oops part.

File: docs/docsite/_ext/cli_manpages/ext.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        path.name.removesuffix('.py')
+        for path in CLI_IMPORTABLES_DIR_PATH.iterdir()
+        if path.name != INIT_MODULE_FILENAME and not path.is_dir()
+    )
+    """List of the importable CLI modules."""
+
+    _jinja2_template = Environment(
+        loader=FileSystemLoader(TEMPLATES_DIR_PATH),
+    ).get_template(CLI_TEMPLATE_FILENAME)
+    """Reusable pre-loaded jinja2 template being rendered in the directive."""
```
Comment (oraNod) [inline docs/docsite/_ext/cli_manpages/ext.py]:
```suggestion
    """Reusable pre-loaded Jinja2 template being rendered in the directive."""
```

File: docs/docsite/_ext/cli_manpages/ext.py (hunk 3 above)
Comment (webknjaz) [reply docs/docsite/_ext/cli_manpages/ext.py]: @oraNod This suggestion exceeds the 72 char limit for the docstring blocks (see PEP 8). It'd need to be reflowed if accepted.

File: test/sanity/code-smell/rstcheck.py (hunk 6 above)
Comment (webknjaz) [reply test/sanity/code-smell/rstcheck.py:117]: I think I'd like to keep this. Unless @mattclay would also want it out. This case is something that should never happen IRL...

File: packaging/pep517_backend/_manpage_generation.py (hunk 5 above)
Comment (webknjaz) [reply packaging/pep517_backend/_manpage_generation.py]: Pulled this in manually. Marking the thread as resolved.

File: docs/docsite/_ext/cli_manpages/ext.py (hunk 7 above)
Comment (webknjaz) [reply docs/docsite/_ext/cli_manpages/ext.py]: Applied this to the corresponding commit. Resolved.

File: docs/docsite/_ext/cli_manpages/ext.py (hunk 4 above)
Comment (webknjaz) [reply docs/docsite/_ext/cli_manpages/ext.py]: Applied this to the corresponding commit. Resolved.

File: docs/docsite/_ext/cli_manpages/ext.py (hunk 3 above)
Comment (webknjaz) [reply docs/docsite/_ext/cli_manpages/ext.py]: Applied this to the corresponding commit with reflowing. Resolved.

File: lib/ansible/cli/console.py (hunk 2 above)
Comment (webknjaz) [reply lib/ansible/cli/console.py]: Marking this thread as resolved.

File: test/sanity/code-smell/rstcheck.py [hunk 8]
```diff
@@ -4,59 +4,170 @@
 import re
 import subprocess
 import sys
+import typing as t
+from shlex import join as _shlex_join
```
Comment (mattclay) [inline test/sanity/code-smell/rstcheck.py:8]: This seems like a lot of unnecessary refactoring and excessive changes for what is actually needed. A much simpler change applies the directive ignore to the relevant files:

```diff
diff --git a/test/sanity/code-smell/rstcheck.py b/test/sanity/code-smell/rstcheck.py
index 4b5a43e69c..6efd2ff85c 100644
--- a/test/sanity/code-smell/rstcheck.py
+++ b/test/sanity/code-smell/rstcheck.py
@@ -20,21 +20,33 @@ def main():
         '-c', 'import rstcheck; rstcheck.main();',
         '--report', 'warning',
         '--ignore-substitutions', ','.join(ignore_substitutions),
-    ] + paths
+    ]
 
-    process = subprocess.run(cmd,
-                             stdin=subprocess.DEVNULL,
-                             stdout=subprocess.PIPE,
-                             stderr=subprocess.PIPE,
-                             check=False,
-                             )
+    docsite_path = 'docs/docsite/rst/cli/'
 
-    if process.stdout:
-        raise Exception(process.stdout)
+    commands = [
+        ['--ignore-directives', 'ansible-cmd-docs'] + [path for path in paths if path.startswith(docsite_path)],
+        [path for path in paths if not path.startswith(docsite_path)]
+    ]
+
+    errors = ''
+
+    for command in commands:
+        process = subprocess.run(cmd + command,
+                                 stdin=subprocess.DEVNULL,
+                                 stdout=subprocess.PIPE,
+                                 stderr=subprocess.PIPE,
+                                 check=False,
+                                 )
+
+        if process.stdout:
+            raise Exception(process.stdout)
+
+        errors += process.stderr.decode(encoding)
 
     pattern = re.compile(r'^(?P<path>[^:]*):(?P<line>[0-9]+): \((?P<level>INFO|WARNING|ERROR|SEVERE)/[0-4]\) (?P<message>.*)$')
 
-    results = parse_to_list_of_dict(pattern, process.stderr.decode(encoding))
+    results = parse_to_list_of_dict(pattern, errors)
 
     for result in results:
         print('%s:%s:%s: %s' % (result['path'], result['line'], 0, result['message']))
```

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/sanity/code-smell/rstcheck.py (hunk 8 above)
Comment (webknjaz) [reply test/sanity/code-smell/rstcheck.py:8]: @mattclay this approach makes it harder to follow and doesn't process some of the failures as gracefully. So I'd like to keep it refactored. Would it be okay if I separated out refactoring into a dedicated PR?

Comment (mattclay) [reply test/sanity/code-smell/rstcheck.py:8]: @webknjaz Can you share example(s) of the kinds of failures you're referring to?

Comment (webknjaz) [reply test/sanity/code-smell/rstcheck.py:8]: Here's my concerns about the original implementation and the patch proposal:
1) An unspecified generic exception being used all over the place makes it hard to reason about the control flow
2) An unobvious structuring side-effect: unexpected output of the first subprocess invocation prevents us from seeing if the following invocation has anything useful to report
3) Unnecessary Python 2 compatibility hacks (working with bytes, old-style string interpolation etc)
4) It is poorly structured (things like a regexp being declared in an arbitrary scope that is not related to using it directly) which manifests as parts being tightly coupled for no good reason — it's just not clean code

Comment (mattclay) [reply test/sanity/code-smell/rstcheck.py:8]: > 1. An unspecified generic exception being used all over the place makes it hard to reason about the control flow

There are only two exceptions raised in the current implementation, both of which terminate the script with an actionable error message. Replacing that with six raised exceptions and three exception handlers makes the code significantly more complex.

> 2. An unobvious structuring side-effect: unexpected output of the first subprocess invocation prevents us from seeing if the following invocation has anything useful to report

I don't consider this an issue worth increasing code complexity over. I can't recall the last time I encountered unexpected output from the subprocess in this test. If it does occur, there's nothing wrong with having to resolve that issue before seeing further output.

> 3. Unnecessary Python 2 compatibility hacks (working with bytes, old-style string interpolation etc)

There's a single use of `bytes.decode` which could be replaced by passing `text=True` to `subprocess.run`. There are two strings that could be updated to use f-strings. There's nothing broken about either of those, and modernizing them only involves removing two lines of code, adding one, and changing one.

> 4. It is poorly structured (things like a regexp being declared in an arbitrary scope that is not related to using it directly) which manifests as parts being tightly coupled for no good reason — it's just not clean code

As a fairly trivial program, there's not much structure that could be reasonably simplified. For the regex example given, that could be handled by moving the regex into the function. Again, it's not broken, and it has a debatable impact on readability.

If there was good reason to refactor the test (I don't think there is), then those changes would be more appropriate in separate PR.

Overall, as a currently very simple sanity test, I think there's much more value in a small, simple implementation  than one nearly three times the size that adds significant complexity. I realize it's a matter of perspective, but I can much more easily understand the current implementation (and my patched version) than the more complex one in this PR.

Comment (webknjaz) [reply test/sanity/code-smell/rstcheck.py:8]: I understand that some of these are perspective-dependent and wouldn't be changing things if I didn't believe this is the right thing to do. Obviously, you can more easily understand your version and I can more easily understand my version. This is also because I've spent some time thinking things through just recently and have a fresh memory of the first impression and things I wanted to improve.

Another bit I noticed missing in your patch (that I recall trying to solve in my refactoring cleanly) is that there's already some structure to extra arguments for `--ignore-substitutions` but not for the new thing. This brings in some new levels of messiness.

Now that I've had a chance to think about this again, I looked into my implementation and remembered another thing that I solved — calling the check against a single document. I know that with your implementation, it'll always fail because you'll be passing an empty list of targets into one of the subcommands which makes it error out.

FWIW as you've started suggesting ways to address my concerns, I got a feeling that you're starting to go through a similar thought process and given enough iterations would end up with something close to what I did. On the high level, at least..

I've already suggested separating the refactoring out into a separate PR specifically because I think there's a good reason for doing this.

Comment (mattclay) [reply test/sanity/code-smell/rstcheck.py:8]: Lets go with a minimal set of test changes for this PR. Any refactoring can be considered in a separate PR later.

I updated my patch (against devel) to account for the one file scenario you mentioned.

```diff
diff --git a/test/sanity/code-smell/rstcheck.py b/test/sanity/code-smell/rstcheck.py
index 4b5a43e69c..13b619c081 100644
--- a/test/sanity/code-smell/rstcheck.py
+++ b/test/sanity/code-smell/rstcheck.py
@@ -20,21 +20,35 @@ def main():
         '-c', 'import rstcheck; rstcheck.main();',
         '--report', 'warning',
         '--ignore-substitutions', ','.join(ignore_substitutions),
-    ] + paths
+    ]
 
-    process = subprocess.run(cmd,
-                             stdin=subprocess.DEVNULL,
-                             stdout=subprocess.PIPE,
-                             stderr=subprocess.PIPE,
-                             check=False,
-                             )
+    docsite_path = 'docs/docsite/rst/cli/'
 
-    if process.stdout:
-        raise Exception(process.stdout)
+    scenarios = [
+        (['--ignore-directives', 'ansible-cmd-docs'], [path for path in paths if path.startswith(docsite_path)]),
+        ([], [path for path in paths if not path.startswith(docsite_path)]),
+    ]
+
+    commands = [cmd + options + files for options, files in scenarios if files]
+
+    errors = ''
+
+    for command in commands:
+        process = subprocess.run(command,
+                                 stdin=subprocess.DEVNULL,
+                                 stdout=subprocess.PIPE,
+                                 stderr=subprocess.PIPE,
+                                 check=False,
+                                 )
+
+        if process.stdout:
+            raise Exception(process.stdout)
+
+        errors += process.stderr.decode(encoding)
 
     pattern = re.compile(r'^(?P<path>[^:]*):(?P<line>[0-9]+): \((?P<level>INFO|WARNING|ERROR|SEVERE)/[0-4]\) (?P<message>.*)$')
 
-    results = parse_to_list_of_dict(pattern, process.stderr.decode(encoding))
+    results = parse_to_list_of_dict(pattern, errors)
 
     for result in results:
         print('%s:%s:%s: %s' % (result['path'], result['line'], 0, result['message']))
```

File: packaging/pep517_backend/_manpage_generation.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+def _make_in_tree_ansible_importable() -> None:
+    """Add the library directory to module lookup paths."""
+    lib_path = str(PROJECT_DIR_PATH / 'lib/')
+    sys.path.insert(0, lib_path)  # NOTE: for the current runtime session
+
+
+@lru_cache(maxsize=1)
+def _make_in_tree_sphinx_extension_importable() -> None:
+    """Add the Sphinx extension directory to module lookup paths."""
+    sphinx_ext_path = str(PROJECT_DIR_PATH / 'docs' / 'docsite' / '_ext')
```
Comment (mattclay) [inline packaging/pep517_backend/_manpage_generation.py:31]: The build backend should be independent of the `docs` directory. At a glance, it looks like that means the Sphinx extension and the templates will need to be moved into the backend.

Comment (webknjaz) [reply packaging/pep517_backend/_manpage_generation.py:31]: @mattclay I wanted to keep the changes small. So yea, there's some dependency on the docs bits that cannot be avoided right now. I thought I'd avoid having copy-pasted duplicate files for now.

Comment (mattclay) [reply packaging/pep517_backend/_manpage_generation.py:31]: I'm less concerned about changes to existing files in the `docs` directory. Almost half of the lines added by this PR are new files there. The Sphinx extension and CLI stubs using it should be somewhere under the `packaging` directory. The docs build process should be able to use them without having to duplicate files in the repository.

This will allow us to remove dependencies on the `docs` directory instead of adding more.

Comment (mattclay) [reply packaging/pep517_backend/_manpage_generation.py:31]: The split of the `docs` directory into https://github.com/ansible/ansible-documentation has started. That directory is effectively frozen in this repository until its removal. The man page builds for releases and docs are now independent of each other, so this PR can focus entirely on changes to the build backend.

**Comment by mattclay** [comment]
Closing, since the docs have been moved and the man pages are no longer included in the sdist. Any use of a Sphinx extension will be in the https://github.com/ansible/ansible-documentation repository.
